challenge_cleanup/src/challenge_cleanup/self_cleanup.py

```diff
@@ -0,0 +1,243 @@
+
+        self._robot = robot
+        self._selected_entity_designator = selected_entity_designator
+
+    def _resolve(self):
+        e = self._selected_entity_designator.resolve()
+
+        item_category = challenge_knowledge.common.get_object_category(e.type)
+        if item_category is not None:
+            location, area_name = challenge_knowledge.common.get_object_category_location(item_category)
+            entities = self._robot.ed.get_entities(id=location)
+            if entities:
+                return entities[0]
+            else:
+                return None
+        else:
+            rospy.logerr("Could not resolve the selected entity!")
+            return None
+
+class storeAreaDesignator(Designator):
+    def __init__(self, robot, name, selected_entity_designator):
+        super(storeAreaDesignator, self).__init__(resolve_type=str, name=name)
+
+        self._robot = robot
+        self._selected_entity_designator = selected_entity_designator
+
+    def _resolve(self):
+        e = self._selected_entity_designator.resolve()
+
+        #return "on_top_of"
+        item_category = challenge_knowledge.common.get_object_category(e.type)
+        if item_category is not None:
+            location, area_name = challenge_knowledge.common.get_object_category_location(item_category)
+            return area_name
+        else:
+            rospy.logerr("Could not resolve the selected entity!")
+            return None
+
+class DetermineCleanupLocation(smach.State):
+    def __init__(self, robot, selected_entity_designator):
+        smach.State.__init__(self, outcomes=["trashbin", "other", "failed"])
+        self._robot = robot
+        self._selected_entity_designator = selected_entity_designator
+
+    def execute(self, userdata):
+        selected_entity = self._selected_entity_designator.resolve()
+
+        if not selected_entity:
+            rospy.logerr("Could not resolve the selected entity!")
+            return "failed"
+
+        rospy.loginfo("The type of the entity is '%s'" % selected_entity.type)
+
+        if challenge_knowledge.common.is_known_object(selected_entity.type):
+            return "other"
+        else:
+            return "trashbin"
+
+class ArmFree(smach.State):
+    def __init__(self, robot):
+        smach.State.__init__(self, outcomes=["yes", "no"])
+        self._robot = robot
+
+    def execute(self, userdata):
+        d = UnoccupiedArmDesignator(self._robot, {}, name="empty_arm_designator2")
+        if d.resolve():
+            return "yes"
+        return "no"
+
+class ArmOccupied(smach.State):
+    def __init__(self, robot):
+        smach.State.__init__(self, outcomes=["yes", "no"])
+        self._robot = robot
+
+    def execute(self, userdata):
+        d = OccupiedArmDesignator(self._robot, {}, name="empty_arm_designator2")
+        if d.resolve():
+            return "yes"
+        return "no"
+
+class Speak(smach.State):
```

Comment:
Because this is code copied from the r5cop-demo. I will need some help if the suggested  conversion is to be implemented. For now, I will leave it as is....